File: core/utils/vertices.js
import { ctx } from "../world/canvas.js";
import { worldToScreen } from "../world/utils.js";

export function drawVertices(vertices, color = "blue",) {
    ctx.save();
    ctx.beginPath()
    ctx.lineWidth = 2;
    ctx.fillStyle = color;
    // ctx.strokeStyle = color;
    ctx.globalAlpha = 0.5;
    ctx.moveTo(vertices[0].x, vertices[0].y);

    vertices.forEach((vertex, i) => {
        if (i) {
            ctx.lineTo(vertex.x, vertex.y);
        }
    });

    ctx.lineTo(vertices[0].x, vertices[0].y);
    ctx.fill();
    // ctx.stroke();
    ctx.restore();
}

export function tranformVertices(vertices, cx, cy, scaleX, scaleY, rotation) {
    const cos = Math.cos(rotation);
    const sin = Math.sin(rotation);

    const hw = scaleX / 2;
    const hh = scaleY / 2

    const result = [];

    for (const vertex of vertices) {
        const sx = vertex.x * hw;
        const sy = vertex.y * hh;

        const rx = sx * cos - sy * sin;
        const ry = sx * sin + sy * cos;

        result.push({
            x: cx + rx,
            y: cy + ry
        });
    }

    return result;
}


export function createVerticesPath(vertices) {
    const verticesPath = new Path2D();

    verticesPath.moveTo(vertices[0].x, vertices[0].y);

    vertices.forEach((vertex, i) => {
        if (i) {
            verticesPath.lineTo(vertex.x, vertex.y);
        }
    });

    verticesPath.lineTo(vertices[0].x, vertices[0].y);

    return verticesPath;
}

export function drawVerticesPath(path, color = "blue", fill = true) {
    ctx.save();
    ctx.beginPath()
    ctx.lineWidth = 2;
    ctx.fillStyle = color;
    // ctx.strokeStyle = color;
    ctx.globalAlpha = 1;

    if (fill)
        ctx.fill(path);
    else
        ctx.stroke(path)
    ctx.restore();
}

export function getVertices(vertices, x, y, width, height, angle) {
    const world = worldToScreen(x, y);
    const transformedVertices = tranformVertices(vertices, world.x, world.y, width, height, angle);

    return transformedVertices;
}

File: void_front/game/core/utils/vertices.py
from __future__ import annotations

import math

import numpy as np

from game.core.world.canvas import draw_polygon
from game.core.world.utils import worldToScreen


def _as_points(vertices):
    points = []
    for vertex in vertices:
        if isinstance(vertex, dict):
            points.append((vertex["x"], vertex["y"]))
        else:
            points.append((vertex.x, vertex.y))
    return np.asarray(points, dtype=float)


def drawVertices(vertices, color="blue"):
    draw_polygon(vertices, color=color, width=0, alpha=128)


def tranformVertices(vertices, cx, cy, scaleX, scaleY, rotation):
    points = _as_points(vertices)
    if not len(points):
        return []
    cos = math.cos(rotation)
    sin = math.sin(rotation)
    hw = scaleX / 2
    hh = scaleY / 2
    scaled = points * np.array([hw, hh], dtype=float)
    rot = np.array([[cos, -sin], [sin, cos]], dtype=float)
    transformed = scaled @ rot.T
    transformed[:, 0] += cx
    transformed[:, 1] += cy
    return [{"x": float(x), "y": float(y)} for x, y in transformed]


def createVerticesPath(vertices):
    return [{"x": v["x"], "y": v["y"]} if isinstance(v, dict) else {"x": v.x, "y": v.y} for v in vertices]


def drawVerticesPath(path, color="blue", fill=True):
    draw_polygon(path, color=color, width=0 if fill else 2, alpha=255)


def getVertices(vertices, x, y, width, height, angle):
    screen = worldToScreen(x, y)
    return tranformVertices(vertices, screen["x"], screen["y"], width, height, angle)
